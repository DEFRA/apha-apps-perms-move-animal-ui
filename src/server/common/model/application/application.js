import { DestinationSection } from '../section/destination/destination.js'
import { LicenceSection } from '../section/licence/licence.js'
import { OriginSection } from '../section/origin/origin.js'
import { validateApplication } from './validation.js'

/**
 * export @typedef {{
 * origin: OriginData | undefined;
 * licence: LicenceData | undefined;
 * destination: DestinationData | undefined;
 * }} ApplicationData
 * @import {OriginData} from '../section/origin/origin.js'
 * @import {LicenceData} from '../section/licence/licence.js'
 * @import {DestinationData} from '../section/destination/destination.js'
 */

export class ApplicationModel {
  _data

  constructor(data) {
    this._data = data
    Object.seal(this)
  }

  validate() {
    return validateApplication(this._data)
  }

  /**
   * Getter for tasks.
   * @returns {Array|Object} The data associated with each and every task in the application
   */
  get tasks() {
    return this._data
  }

  /**
   * @param {ApplicationData | undefined} state
   * @returns {ApplicationModel}
   */
  static fromState(state) {
    return new ApplicationModel({
      origin: OriginSection.fromState(state?.origin),
      licence: LicenceSection.fromState(state?.licence),
      destination: DestinationSection.fromState(state?.destination)
    })
  }
}
