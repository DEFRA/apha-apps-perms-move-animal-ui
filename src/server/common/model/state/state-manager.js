/** @import {Request} from "@hapi/hapi/lib/types/request.js" */

/**
 * @typedef {Record<string, any>} RawSectionState
 * @typedef {Record<string, RawSectionState>} RawApplicationState
 */

const sectionKeys = ['origin', 'destination', 'licence', 'biosecurity']

export class StateManager {
  /** @param {Request} request */
  constructor(request) {
    this._request = request
  }

  /** @returns {RawApplicationState} */
  toState() {
    return Object.fromEntries(
      sectionKeys.map((key) => [key, this._request.yar.get(key)])
    )
  }
}
