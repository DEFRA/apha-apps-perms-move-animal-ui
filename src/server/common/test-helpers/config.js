import { config } from '~/src/config/config.js'

/* global jest */

const configGet = config.get.bind(config)

/**
 * @param {string} key
 * @param {any} value
 */
export const spyOnConfig = (key, value) => {
  jest.spyOn(config, 'get').mockImplementation((name) => {
    if (name === key) {
      return value
    } else {
      return configGet(name)
    }
  })
  return config
}
