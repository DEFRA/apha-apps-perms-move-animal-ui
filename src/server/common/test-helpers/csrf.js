import { statusCodes } from '~/src/server/common/constants/status-codes.js'
/* global expect, test */

/**
 * @param {OutgoingHttpHeaders} headers
 * @param {string} name
 * @returns {string}
 */
const findCookie = (headers, name) => {
  const setCookie = headers['set-cookie']

  /** @type {string[] | undefined} */
  const cookies = typeof setCookie === 'string' ? [setCookie] : setCookie
  return cookies?.find((h) => h?.includes(`${name}=`)) ?? ''
}

/**
 * @param {string} cookieString
 */
const extractCookieValue = (cookieString) =>
  cookieString.split('=')[1].split(';')[0]

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedPost = (serverFn, injectionOptions) =>
  test('should reject if experiencing an apparent csrf attack (they have their session cookie, but no hidden form value)', async () => {
    const { statusCode } = await serverFn().inject({
      ...injectionOptions,
      headers: {
        ...(injectionOptions.headers ?? {}),
        Cookie: 'crumb=csrf-value'
      }
    })

    expect(statusCode).toEqual(statusCodes.forbidden)
  })

/**
 * @param {function(): Server} serverFn
 * @param {ServerInjectOptions} injectionOptions
 */
export const testCsrfProtectedGet = (serverFn, injectionOptions) =>
  test('should contain a cookie with a set CSRF value, and contain a hidden form field to submit that CSRF back to the server', async () => {
    const { headers, payload } = await serverFn().inject(injectionOptions)

    const token = extractCookieValue(findCookie(headers, 'crumb'))
    expect(token).not.toHaveLength(0)
    expect(payload).toContain(
      `<input type="hidden" name="crumb" value="${token}"`
    )
  })

/**
 * @param {ServerInjectOptions} injectionOptions
 * @returns {ServerInjectOptions}
 */
export const withCsrfProtection = (injectionOptions) => ({
  ...injectionOptions,
  payload: {
    ...(typeof injectionOptions.payload === 'object'
      ? injectionOptions.payload
      : {}),
    crumb: 'csrf-value'
  },
  headers: {
    ...(injectionOptions.headers ?? {}),
    Cookie: 'crumb=csrf-value'
  }
})

/**
 * @import { OutgoingHttpHeaders } from 'http'
 * @import { ServerInjectOptions, Server } from '@hapi/hapi'
 */
