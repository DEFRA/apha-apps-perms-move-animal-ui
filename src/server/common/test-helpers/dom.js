import { JSDOM } from 'jsdom'

/**
 * @param {string} payload
 * @returns {Document}
 */
export const parseDocument = (payload) => new JSDOM(payload).window.document
