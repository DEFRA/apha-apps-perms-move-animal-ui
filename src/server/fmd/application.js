import { ApplicationModel } from '../common/model/application/application.js'
import { AboutSection } from './about/section.js'
import { DestinationSection } from './destination/section.js'
import { DisposalSection } from './disposal/section.js'
import { LicenceSection } from './licence/section.js'
import { MovementDetailsSection } from './movement-details/section.js'
import { OriginSection } from './origin/section.js'
import { SlaughterInformationSection } from './slaughter/section.js'

/**
 * @import { SectionModel } from '../common/model/section/section-model/section-model.js'
 */

export class FmdApplicationModel extends ApplicationModel {
  /** @type {typeof SectionModel[]} */
  static implementedSections = [
    AboutSection,
    DestinationSection,
    OriginSection,
    MovementDetailsSection,
    LicenceSection,
    SlaughterInformationSection,
    DisposalSection
  ]

  get version() {
    return {
      major: 1,
      minor: 0
    }
  }

  get journeyId() {
    return 'GET_PERMISSION_TO_MOVE_ANIMALS_UNDER_DISEASE_CONTROLS_FMD'
  }
}
