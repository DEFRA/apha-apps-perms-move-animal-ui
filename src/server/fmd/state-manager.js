import { StateManager } from '../common/model/state/state-manager.js'

export class FmdStateManager extends StateManager {
  get key() {
    return FmdStateManager.key
  }

  static get key() {
    return 'fmd-application'
  }
}
