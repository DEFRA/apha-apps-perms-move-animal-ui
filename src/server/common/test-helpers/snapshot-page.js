import { parseDocument } from '../../common/test-helpers/dom.js'
import { createServer } from '~/src/server/index.js'
import { withCsrfProtection } from './csrf.js'
/** @import { Server } from '@hapi/hapi' */

/* global expect, it, beforeAll, afterAll, describe */

/**
 * @param {string} describeMessage
 * @param {string} shouldMessage
 * @param {string} urlPath
 */
export const describePageSnapshot = (
  describeMessage,
  shouldMessage,
  urlPath
) => {
  describe(describeMessage, () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it(shouldMessage, async () => {
      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: urlPath
        })
      )

      const content =
        parseDocument(payload).querySelector('#main-content')?.innerHTML

      expect(content).toMatchSnapshot()
    })
  })
}
