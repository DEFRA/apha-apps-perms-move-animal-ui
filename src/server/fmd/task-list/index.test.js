import { describePageSnapshot } from '../../common/test-helpers/snapshot-page.js'
import { FmdApplicationModel } from '../application.js'
import { FmdStateManager } from '../state-manager.js'
import { FmdTaskListController } from './index.js'
import { createServer } from '../../index.js'
import { withCsrfProtection } from '../../common/test-helpers/csrf.js'
import { parseDocument } from '../../common/test-helpers/dom.js'
import { spyOnConfig } from '../../common/test-helpers/config.js'

/** @import { Server } from '@hapi/hapi' */

describe('FmdTaskListController', () => {
  const controller = new FmdTaskListController()

  it('should have the right title', () => {
    expect(controller.pageTitleAndHeading).toBe(
      'Your foot and mouth disease movement licence application'
    )
  })

  it('should contain the right application model', () => {
    expect(controller.ApplicationModel).toBe(FmdApplicationModel)
  })

  it('should contain the state manager', () => {
    expect(controller.StateManager).toBe(FmdStateManager)
  })

  it('should have the right url path', () => {
    expect(controller.urlPath).toBe('/fmd/task-list')
  })

  it('should have the right submit url path', () => {
    expect(controller.submitUrlPath).toBe('/fmd/submit/check-answers')
  })

  describePageSnapshot({
    describes: 'FmdTaskListController',
    it: 'should have the expected content',
    pageUrl: controller.urlPath
  })

  describe('task list banner', () => {
    /** @type {Server} */
    let server

    beforeAll(async () => {
      server = await createServer()
      await server.initialize()
    })

    afterAll(async () => {
      await server.stop({ timeout: 0 })
    })

    it('should have the right banner (in protype mode)', async () => {
      spyOnConfig('featureFlags', {
        prototypeMode: true
      })

      const { payload } = await server.inject(
        withCsrfProtection({
          method: 'GET',
          url: controller.urlPath
        })
      )

      const contentTag = parseDocument(payload)
        .querySelector('.govuk-phase-banner__content__tag')
        ?.innerHTML.trim()
      const text = parseDocument(payload)
        .querySelector('.govuk-phase-banner__text')
        ?.innerHTML.trim()

      expect(contentTag).toBe('Prototype')
      expect(text).toBe('This is a service prototype.')
    })
  })
})
