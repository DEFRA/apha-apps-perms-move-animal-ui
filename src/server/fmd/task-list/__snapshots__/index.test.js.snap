// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FmdTaskListController FmdTaskListController should have the expected content 1`] = `
"
  <div class="govuk-grid-row">
    <div class="govuk-grid-column-two-thirds">
      <div class="govuk-body" data-testid="app-page-body">
        <h1 class="govuk-heading-l">Your foot and mouth disease movement licence application</h1>
          <p>
            You have
            <span data="test-id-task-to-complete-count">2 out of 2</span>
            sections left to complete.
          </p>
        
<ul class="govuk-task-list">
  <li class="govuk-task-list__item govuk-task-list__item--with-link">
    <div class="govuk-task-list__name-and-hint">
      <a class="govuk-link govuk-task-list__link" href="/fmd/about-the-movement-or-activity/type" aria-describedby="task-list-1-status">
        About the movement
      </a>
    </div>
    <div class="govuk-task-list__status" id="task-list-1-status">
      <strong class="govuk-tag govuk-tag--blue">
        Incomplete
      </strong>
    </div>
  </li>
  <li class="govuk-task-list__item">
    <div class="govuk-task-list__name-and-hint">
      <div>
        Receiving the licence
      </div>
    </div>
    <div class="govuk-task-list__status govuk-task-list__status--cannot-start-yet" id="task-list-2-status">
      Cannot start yet
    </div>
  </li>
</ul>

        <form method="POST">
          <input type="hidden" name="crumb" value="csrf-value"> <button type="submit" class="govuk-button govuk-button--secondary" data-module="govuk-button">
  Review and submit
</button>
        </form>
      </div>
    </div>
  </div>
        "
`;
